WangscapeGui/mainwindow.cpp

```diff
@@ -0,0 +1,202 @@
+        return;
+    }
+
+    resetTilesetGenerator();
+
+    mUi->comboBox->clear();
+    mPreviewImages.clear();
+
+    mTilesetGenerator->generate([this](const sf::Texture& output, std::string filename)
+    {
+        const sf::Image source_image = output.copyToImage();
+
+        const auto basename = filename.substr(filename.rfind('/') + 1);
+        const auto qimage = convertSfImageToQImage(source_image);
+        mPreviewImages.emplace(std::make_pair(basename,
+                                             std::make_pair(filename, std::move(qimage))));
+    });
+
+    initializePreviewArea();
+    displayTilesetPreview();
+}
+
+void MainWindow::saveOutput()
+{
+    if (mOptionsEditor->useDefaultOuputDir())
+    {
+        mOptions.relativeOutputDirectory = mOriginalOptions.relativeOutputDirectory;
+    }
+    else
+    {
+        QString chosen_path = QFileDialog::getExistingDirectory(this, tr("Select output dir"));
+        mOptions.relativeOutputDirectory = chosen_path.toLocal8Bit().constData();
+    }
+
+    if (mPreviewImages.empty())
+    {
+        QMessageBox warning_message;
+        warning_message.warning(this, "Warning", QString("You have to generate tileset before saving it"));
+        return;
+    }
+
+    // set to mOptions.relativeOutputDirectory not part of preview_image filepath,
+    // cause it is is used to generate that filepath and it makes
+    // modifying output dir for the whole output a lot easier
+    const std::string image_dir = mOptions.relativeOutputDirectory;
+    boost::filesystem::create_directories(image_dir);
+    boost::filesystem::path p(image_dir);
+    for (const auto& preview_image : mPreviewImages)
+    {
+        p.append(preview_image.first);
+        if (!preview_image.second.second.save(QString(p.string().c_str())))
+        {
+            QMessageBox critical_error_message;
+            critical_error_message.critical(this, "Error during saving tilesets", QString("Couldn't save tileset image"));
+            return;
+        }
+        p.remove_filename();
+    }
+
+    mTilesetGenerator->metaOutput.writeAll(mOptions);
+
+    QMessageBox success_message;
+    success_message.information(this, "Saved!", QString("Output has been saved successfully"));
+}
+
+void MainWindow::resetOptions()
+{
+    mOptions = mOriginalOptions;
+    mOptionsEditor->setOptions(&mOptions);
+}
+
+void MainWindow::loadOptionsFromFile()
+{
+    QString chosen_path = QFileDialog::getOpenFileName(this,
+                                                     tr("Select options file"));
+    if (chosen_path.isEmpty())
+        return;
+
+    mUi->currentOptionsPath->setText(QString(chosen_path));
+
+    // TODO(hryniuk): make OptionsManager throw custom exception, catch it here
+    // and show QMessageBox with a proper error message
+    mOptionsManager = std::make_unique<OptionsManager>(chosen_path.toLocal8Bit().constData());
```

Comment:
We have a global `OptionsManager` instance now; is it possible to use that instead?